File: 03-Binary-Search/02-Binary-Search-On-Answer/cpp/koko_eating_bananas.cpp
/*
 * Problem: Koko Eating Bananas
 * Source:  LeetCode 875 - https://leetcode.com/problems/koko-eating-bananas/
 * 
 * Approach: Binary search on the answer (eating speed k).
 *           For each k, check if Koko can finish all piles in h hours.
 * Time Complexity:  O(n * log(max_pile))
 * Space Complexity: O(1)
 * Tags: binary-search-on-answer
 */

#include <bits/stdc++.h>
using namespace std;

bool canFinish(vector<int>& piles, int k, int h) {
    long long hours = 0;
    for (int p : piles)
        hours += (p + k - 1) / k;  // ceil division
    return hours <= h;
}

int minEatingSpeed(vector<int>& piles, int h) {
    int lo = 1, hi = *max_element(piles.begin(), piles.end());
    while (lo < hi) {
        int mid = lo + (hi - lo) / 2;
        if (canFinish(piles, mid, h)) hi = mid;
        else lo = mid + 1;
    }
    return lo;
}

int main() {
    vector<int> piles = {3, 6, 7, 11};
    cout << "Min speed (h=8): " << minEatingSpeed(piles, 8) << endl;  // 4
    
    vector<int> piles2 = {30, 11, 23, 4, 20};
    cout << "Min speed (h=5): " << minEatingSpeed(piles2, 5) << endl; // 30
    cout << "Min speed (h=6): " << minEatingSpeed(piles2, 6) << endl; // 23
    return 0;
}

File: 03-Binary-Search/02-Binary-Search-On-Answer/java/KokoEatingBananas.java
/*
 * Problem: Koko Eating Bananas
 * Source:  LeetCode 875
 * Tags: binary-search-on-answer
 */

import java.util.*;

public class KokoEatingBananas {
    
    static boolean canFinish(int[] piles, int k, int h) {
        long hours = 0;
        for (int p : piles)
            hours += (p + k - 1) / k;
        return hours <= h;
    }
    
    public static int minEatingSpeed(int[] piles, int h) {
        int lo = 1, hi = Arrays.stream(piles).max().getAsInt();
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (canFinish(piles, mid, h)) hi = mid;
            else lo = mid + 1;
        }
        return lo;
    }
    
    public static void main(String[] args) {
        System.out.println("Min speed (h=8): " + minEatingSpeed(new int[]{3, 6, 7, 11}, 8)); // 4
        System.out.println("Min speed (h=5): " + minEatingSpeed(new int[]{30, 11, 23, 4, 20}, 5)); // 30
    }
}
